internal/journal.go

```diff
@@ -8,163 +8,304 @@ import (
+
+		// We should update previous JournalInfo to zero value if there are no JournalInfo after that
+		prevJi, err := ji.GetPrevious(folder)
+		if err != nil {
+			if err.Error() != cantFindJournal {
+				return err
+			}
+			return nil
+		}
+
+		prevJi.JournalSize = 0
+		return prevJi.Upload(folder)
 	}
+	tracelog.InfoLogger.Printf("found the next journal %+v", nextJi)
 
-	journalSize, err := GetJournalSizeInSemiInterval(
-		folder,
-		journalPath,
-		journalCmpLess,
-		lastSentinel.JournalEnd,
-		newJournalEnd,
-	)
+	nextJi.JournalStart = ji.JournalStart
+	err = nextJi.Upload(folder)
 	if err != nil {
-		tracelog.ErrorLogger.Printf("can not evaluate journal sum for %s: %s", lastSentinelName, err)
 		return err
 	}
-	tracelog.InfoLogger.Printf(
-		"journal size for %s in the semi interval (%s; %s] is equal to %d",
-		lastSentinelName,
-		lastSentinel.JournalEnd,
-		newJournalEnd,
-		journalSize,
-	)
+	tracelog.InfoLogger.Printf("the next journal updated %+v", nextJi)
 
-	err = UploadBackupInfo(folder, lastSentinelName, JournalInfo{
-		JournalStart: lastSentinel.JournalStart,
-		JournalEnd:   lastSentinel.JournalEnd,
-		JournalSize:  journalSize,
-	})
+	err = nextJi.Calculate(folder)
 	if err != nil {
-		tracelog.ErrorLogger.Printf("can not update journal info for %s: %s", lastSentinelName, err)
 		return err
 	}
 
-	tracelog.InfoLogger.Printf("journal info has been updated for %s", lastSentinelName)
 	return nil
 }
 
-// (start;end]
-func GetJournalSizeInSemiInterval(
+// GetLastJournalInfo receives the most recently created JournalInfo on S3
+func GetLastJournalInfo(
 	folder storage.Folder,
-	journalPath string,
-	journalCmpLess func(a, b string) bool,
-	start, end string,
-) (int64, error) {
-	folder = folder.GetSubFolder(journalPath)
-	journalFiles, _, err := folder.ListFolder()
+	journalDir string,
+	journalLessComparator func(a, b string) bool,
+) (JournalInfo, error) {
+	objs, _, err := folder.GetSubFolder(utility.BaseBackupPath).ListFolder()
 	if err != nil {
-		return 0, err
+		return JournalInfo{}, err
+	}
+	if len(objs) == 0 {
+		return JournalInfo{}, nil
+	}
+
+	var lastJournalInfo storage.Object
+	for _, v := range objs {
+		if strings.HasPrefix(v.GetName(), JournalPrefix) &&
+			(lastJournalInfo == nil || journalLessComparator(lastJournalInfo.GetName(), v.GetName())) {
+			lastJournalInfo = v
+		}
+	}
+
+	if lastJournalInfo == nil {
+		return JournalInfo{}, xerrors.New("there are no journals on the S3")
+	}
+
+	lastBackupName := strings.TrimPrefix(lastJournalInfo.GetName(), JournalPrefix)
+	backupInfo, err := NewJournalInfo(
+		lastBackupName,
+		folder,
+		journalDir,
+		journalLessComparator,
+	)
+	if err != nil {
+		return JournalInfo{}, err
+	}
+
+	return backupInfo, nil
+}
+
+// Calculate calculates the size of the JournalInfo in the semi-interval (JournalStart; JournalEnd]
+// using journal files on JournalDirectoryName and save it for the previous JournalInfo
+func (ji *JournalInfo) Calculate(folder storage.Folder) error {
```

Comment:
'Calculate' name doesn't tell much about its behaviour =/ Originally I thought that it is pure function.